cassandra/policies.py

```diff
@@ -864,6 +874,378 @@ def _add_jitter(self, value):
         return min(max(self.base_delay, delay), self.max_delay)
 
 
+class ShardConnectionScheduler(ABC):
+    """
+    A base class for a scheduler for a shard connection backoff policy.
+    ``ShardConnectionScheduler`` is a per Session instance that schedules per shard connections according to
+       ``ShardConnectionBackoffPolicy`` that instantiates it.
+    """
+
+    @abstractmethod
+    def schedule(
+            self,
+            host_id: str,
+            shard_id: int,
+            method: Callable[..., None],
+            *args: List[Any],
+            **kwargs: dict[Any, Any]) -> None:
+        """
+        Schedules request to create connection to given host and shard according to the policy.
+        At no point request is executed on the call, it is always running in a separate thread,
+        this method is non-blocking in this regard.
+
+        ``host_id`` - an id of the host of the shard.
+        ``shard_id`` - an id of the shard.
+        ``method`` - a callable that creates connection and stores it in the connection pool.
+          Currently, it is `HostConnection._open_connection_to_missing_shard`.
+        ``*args`` and ``**kwargs`` are passed to ``method`` when policy executes it.
+        """
+        raise NotImplementedError()
+
+
+class ShardConnectionBackoffPolicy(ABC):
+    """
+    Base class for shard connection backoff policies.
+    These policies allow user to control pace of establishing new connections.
+
+    On `new_scheduler` instantiate a scheduler that behaves according to the policy.
+    """
+
+    @abstractmethod
+    def new_scheduler(self, session: Session) -> ShardConnectionScheduler:
+        raise NotImplementedError()
+
+
+class NoDelayShardConnectionBackoffPolicy(ShardConnectionBackoffPolicy):
+    """
+    A shard connection backoff policy with no delay between attempts and no concurrency restrictions.
+    Ensures that at most one pending connection per (host, shard) pair.
+    If connection attempts for the same (host, shard) it is silently dropped.
+
+    On `new_scheduler` instantiate a scheduler that behaves according to the policy.
+    """
+
+    def new_scheduler(self, session: Session) -> ShardConnectionScheduler:
+        return _NoDelayShardConnectionBackoffScheduler(session)
+
+
+class _NoDelayShardConnectionBackoffScheduler(ShardConnectionScheduler):
+    """
+    A scheduler for ``cassandra.policies.NoDelayShardConnectionBackoffPolicy``.
+    It does not introduce any delay or concurrency restrictions.
+    It only ensures that there is only one pending or scheduled connection per (host, shard) pair.
+    """
+    session: Session
+    already_scheduled: dict[str, bool]
+    lock: threading.Lock
+
+    def __init__(self, session: Session):
+        self.session = weakref.proxy(session)
+        self.already_scheduled = {}
+        self.lock = threading.Lock()
+
```

Comment:
`self.session = weakref.proxy(session)` is another argument for both more descriptive interface docs, and for not using `Session` here.
How is the dev implementing such policy after reading its doc comment supposed to know that they should use `weakref.proxy` here?